### PR TITLE
feat: Ent 5149: added support for programs inside the course.

### PR DESCRIPTION
**Pull request by HammadAhmadWaqas**
## Description

Updated `_update_full_content_metadata_course` task to create course associated program's ContentMedataData records
and updated `index_content_keys_in_algolia` task to index these programs as well.  

## Ticket Link

Link to the associated ticket
[Link title](https://openedx.atlassian.net/browse/ENT-5149)

## Post-review

Squash commits into discrete sets of changes

### CODE REVIEW

File: enterprise_catalog/apps/catalog/models.py [hunk 1]
```diff
@@ -628,6 +647,31 @@ def associate_content_metadata_with_query(metadata, catalog_query):
     associated_content_keys = [metadata.content_key for metadata in metadata_list]
     return associated_content_keys
 
+def create_course_associated_programs(programs, course_content_matadata):
+    """
+    Creates or updates a ContentMetadata object for each entry in `metadata`,
+    and then associates that object with the `content_matadata` provided.
+
+    Arguments:
```
Comment (irfanuddinahmad) [inline enterprise_catalog/apps/catalog/models.py:656]: Are these arguments correct?

File: enterprise_catalog/apps/catalog/models.py [hunk 2]
```diff
@@ -400,9 +400,13 @@ class ContentMetadata(TimeStampedModel):
         null=True,
         db_index=True,
         help_text=_(
-            "The key that represents this content's parent, such as a course or program."
+            "The key represents this content's parent. For example for course_runs content their parent course key."
         )
     )
+
+    # one course can be associate with many programs and one program can contain many courses.
```
Comment (irfanuddinahmad) [inline enterprise_catalog/apps/catalog/models.py]: nit: associated

File: enterprise_catalog/apps/catalog/models.py (hunk 1 above)
Comment (HammadAhmadWaqas) [reply enterprise_catalog/apps/catalog/models.py:656]: fixed typos

File: enterprise_catalog/apps/catalog/models.py (hunk 2 above)
Comment (HammadAhmadWaqas) [reply enterprise_catalog/apps/catalog/models.py]: done

**Review by irfanuddinahmad — APPROVED** [review]
LGTM! ... awesome work!

**Review by sameenfatima78 — APPROVED** [review]
Awesome work!

File: enterprise_catalog/apps/api/tasks.py [hunk 3]
```diff
@@ -505,7 +518,14 @@ def index_content_keys_in_algolia(content_keys, algolia_client):
             customer_uuids_by_key[content_key].update(enterprise_customer_uuids)
             catalog_queries_by_key[content_key].update(enterprise_catalog_queries)
 
-        # iterate through only the courses, retrieving the enterprise-related uuids from the
+            # course metadata might have associated programs. add them as well.
```
Comment (saleem-latif) [inline enterprise_catalog/apps/api/tasks.py:521]: I think this should only be done for programs, not all content types

File: enterprise_catalog/apps/api/tasks.py [hunk 4]
```diff
@@ -468,14 +471,24 @@ def index_content_keys_in_algolia(content_keys, algolia_client):
         catalog_queries_by_key = defaultdict(set)
 
         # retrieve ContentMetadata records that match the specified content_keys in the
-        # content_key or parent_content_key. returns courses, programs and course runs.
-        query = Q(content_key__in=content_keys_batch) | Q(parent_content_key__in=content_keys_batch)
+        # content_key or parent_content_key or course associated programs. returns courses, programs and course runs.
+        query = (
+            Q(content_key__in=content_keys_batch)
+            | Q(parent_content_key__in=content_keys_batch)
+            | Q(
```
Comment (saleem-latif) [inline enterprise_catalog/apps/api/tasks.py:478]: I do not think we need to change the query, It seems old query would give us the correct results. I think the issue that is appearing is being caused by this query update.

**Review by saleem-latif — COMMENTED** [review]
@irfanuddinahmad Added comments in places that seem to the victim. However, debugging will be needed to be sure.

**Pull request merged**